Finfotech pl seller pre prod logs for verification

sudhakara11 opened the pull request.

JyotiSharma0711 commented:
- **Flow 1, Flow 2, Flow 3**

- **on_search**
  - `"SETTLEMENT_TERMS_[1], Term code 'COURT_JURISDICTION' is not present in tag.list"`
  - `"SETTLEMENT_TERMS_[1], List item[0] has an invalid duration value for SETTLEMENT_WINDOW"`

- **on_select_2, on_select_3**
  - `"breakupTotalMismatch": "Total of quote breakup (442302) does not match with price.value (442300)"`
  - `List item descriptor code 'COOL_OF_PERIOD' at index 12 within tag 0 is invalid`
  - `List item descriptor code 'INSTALLEMNT_AMOUNT' at index 13 within tag 0 is invalid`

- **on_init_1, on_init_3**
  - `Term code 'COURT_JURISDICTION' is not present in tag.list`
  - `"SETTLEMENT_TERMS_[1], List item[0] has an invalid duration value for SETTLEMENT_WINDOW"`
  - `Total of quote breakup (442801) does not match with price.value (442300)`

- **on_init_2**
  - `"item0_xinput_head_descriptor": "descriptor.name should be E-mandate at items[0].xinput.head"`

- **on_init_3**
  - `"SETTLEMENT_TERMS_[1], Term code 'COURT_JURISDICTION' is not present in tag.list"`
  - `"SETTLEMENT_TERMS_[1], List item[0] has an invalid duration value for SETTLEMENT_WINDOW"`

- **on_confirm**
  - `"breakupTotalMismatch": "Total of quote breakup (442302) does not match with price.value (442300)"`
  - `"documents[0].mime_type": "Invalid mime_type: text/html"`

sudhakara11 commented:
@JyotiSharma0711 

We have fixed the issues and submitted the logs with same PR

**"SETTLEMENT_TERMS_[1], Term code 'COURT_JURISDICTION' is not present in tag.list"** - These details are available as per developer guide specifications.

https://github.com/ONDC-Official/verification-logs/pull/1908

Regards,
Sudhakar G

@RonitKumarSinha

sourabhsaini11 commented:
@sudhakara11
common_issues
- Total of quote breakup (10017) does not match with price.value (10000)

on_search
- missing tag 'COURT_JURISDICTION' at SETTLEMENT_TERMS, invalid tag COURT_JURIDICTION

on_select
- invalid list code KFS_LINK

on_init
- invalid list code KFS_LINK
- missing tag 'COURT_JURISDICTION' at SETTLEMENT_TERMS, invalid tag COURT_JURIDICTION

on_confirm
- invalid list code KFS_LINK
- missing tag 'COURT_JURISDICTION', 'DELAY_INTEREST', 'SETTLEMENT_TYPE' at SETTLEMENT_TERMS, invalid tag COURT_JURIDICTION

confirm
- missing tags 'SETTLEMENT_WINDOW', 'SETTLEMENT_BASIS', 'MANDATORY_ARBITRATION', 'COURT_JURISDICTION' at SETTLEMENT_TERMS

PrePart
on_update 
- missing PRE_PAYMENT_CHARGE at quote.breakup
- additional INSTALLMENT obj present at on_update_4
- OUTSTANDING_PRINCIPAL value is changed from on_update_4 to on_update_3

Missed Emi
on_update 
- missing LATE_FEE_AMOUNT at quote.breakup
- MISSED_EMI_PAYMENT obj status should be NOT-PAID in on_update_3
- OUTSTANDING_PRINCIPAL value is changed from on_update_4 to on_update_3

foreclosure
on_update 
- missing FORCLOSUER_CHARGES at quote.breakup
- additional INSTALLMENT obj present at on_update_4

sudhakara11 commented:
Dear @sourabhsaini11 

Greetings!

We have submitted the logs with the fixes, kindly review and confirm.

Same PR link

Regards,
Sudhakar G

@RonitKumarSinha

JyotiSharma0711 commented:
### **Flow 1**  
- **on_search:**  
  - `"paymentTags": "SETTLEMENT_TERMS_[1], List item[0] has an invalid duration value for SETTLEMENT_WINDOW"`

- **on_select_3:**  
  - `"tags":`  
    - `"List item descriptor code 'COOL_OF_PERIOD' at index 12 within tag 0 is invalid"`  
    - `"List item descriptor code 'INSTALLEMNT_AMOUNT' at index 13 within tag 0 is invalid"`

- **on_init_1:**  
  - `"tags":`  
    - `"SETTLEMENT_TERMS_[1], List item[0] has an invalid duration value for SETTLEMENT_WINDOW"`
  - `"cancellationFee": "Invalid Cancellation Term[1] - Either external_ref or fulfillment_state with valid cancellation_fee is required"`

- **on_init_2:**  
  - `"tags":`  
    - `"SETTLEMENT_TERMS_[1], List item[0] has an invalid duration value for SETTLEMENT_WINDOW"`

- **on_init_3:**  
  - `"tags":`  
    - `"SETTLEMENT_TERMS_[1], List item[0] has an invalid duration value for SETTLEMENT_WINDOW"`
  - `"cancellationFee": "Invalid Cancellation Term[1] - Either external_ref or fulfillment_state with valid cancellation_fee is required"`

- **on_confirm:**  
  - `"tags":`  
    - `"SETTLEMENT_TERMS_[1], List item[0] has an invalid duration value for SETTLEMENT_WINDOW"`
  - `"cancellationFee": "Invalid Cancellation Term[1] - Either external_ref or fulfillment_state with valid cancellation_fee is required"`

sudhakara11 commented:
Dear @JyotiSharma0711 

Greetings!

We have reviewed the feedback and below are our response 

Flow 1
on_search:

"paymentTags": "SETTLEMENT_TERMS_[1], List item[0] has an invalid duration value for SETTLEMENT_WINDOW" - 
**_P10D (Valid i1806 format)_**

on_select_3:

"tags":
"List item descriptor code 'COOL_OF_PERIOD' at index 12 within tag 0 is invalid" -
 **_PT5M (Valid i1806 format) - As the pramaan is waiting for 5min - we have changed it to 5 minutes - in real time it will be days_**
"List item descriptor code 'INSTALLEMNT_AMOUNT' at index 13 within tag 0 is invalid" - 
**_The installment amount is correct 1080_**

on_init_1:

"tags":
"SETTLEMENT_TERMS_[1], List item[0] has an invalid duration value for SETTLEMENT_WINDOW" - 
**_P10D (Valid i1806 format)_**
"cancellationFee": "Invalid Cancellation Term[1] - Either external_ref or fulfillment_state with valid cancellation_fee is required" - **_0% (0 percent is lender choice, lender wont charge anything as cancellation fee)_**

on_init_2:

"tags":
"SETTLEMENT_TERMS_[1], List item[0] has an invalid duration value for SETTLEMENT_WINDOW" - 
**_P10D (Valid i1806 format)_**

on_init_3:_**

"tags":
"SETTLEMENT_TERMS_[1], List item[0] has an invalid duration value for SETTLEMENT_WINDOW" - 
**_P10D (Valid i1806 format)_**

"cancellationFee": "Invalid Cancellation Term[1] - Either external_ref or fulfillment_state with valid cancellation_fee is required" - **_0% (0 percent is lender choice, lender wont charge anything as cancellation fee)_**

on_confirm:

"tags":
"SETTLEMENT_TERMS_[1], List item[0] has an invalid duration value for SETTLEMENT_WINDOW" - 
**_P10D (Valid i1806 format)_**
"cancellationFee": "Invalid Cancellation Term[1] - Either external_ref or fulfillment_state with valid cancellation_fee is required" - **_0% (0 percent is lender choice, lender wont charge anything as cancellation fee)_**

Regards,
Sudhakar G

sudhakara11 commented:
Dear @JyotiSharma0711 

Greetings!

We have added IGM logs as well in the same PR, Kindly review and confirm.

Regards,
Sudhakar G

JyotiSharma0711 commented:
@sudhakara11 Please submit only the payload of API calls and exclude the request and response details

JyotiSharma0711 commented:
### **on_search**  
- Invalid duration value for `SETTLEMENT_WINDOW` in `SETTLEMENT_TERMS_[1]`.  

### **on_select_2**  
- Invalid descriptor code `COOL_OF_PERIOD` at index 12.  
- Invalid descriptor code `INSTALLEMNT_AMOUNT` at index 13.  
- Invalid descriptor code `KFS_LINK` at index 14.  

### **on_select_3**  
- Invalid descriptor code `COOL_OF_PERIOD` at index 12.  
- Invalid list code `COOL_OF_PERIOD`.  
- Invalid descriptor code `INSTALLEMNT_AMOUNT` at index 13.  
- Invalid list code `INSTALLEMNT_AMOUNT`.  
- Invalid descriptor code `KFS_LINK` at index 14.  
- Invalid list code `KFS_LINK`.  

### **on_init_1**  
- Invalid duration value for `SETTLEMENT_WINDOW` in `SETTLEMENT_TERMS_[1]`.  
- Invalid cancellation term: Either `external_ref` or `fulfillment_state` with valid `cancellation_fee` is required.  

### **on_init_2**  
- Same issues as `on_init_1`.  

### **on_init_3**  
- Same issues as `on_init_1`.  

### **confirm**  
- `SETTLEMENT_TERMS_[1]` is missing:  
  - `SETTLEMENT_WINDOW`  
  - `SETTLEMENT_BASIS`  
  - `MANDATORY_ARBITRATION`  
  - `COURT_JURISDICTION`  

### **on_confirm**  
- Invalid duration value for `SETTLEMENT_WINDOW` in `SETTLEMENT_TERMS_[1]`.  
- Invalid cancellation term: Either `external_ref` or `fulfillment_state` with valid `cancellation_fee` is required.

sudhakara11 commented:
@sourabhsaini11 

As requested - we have fixed the comments and submitted the latest logs for the verification

Kindly review the transaction and IGM logs and provide the feedback

Regards,
Sudhakar G

JyotiSharma0711 commented:
Hi @sudhakara11 

on_select_3
Invalid descriptor code COOL_OF_PERIOD at index 12.
Refer to this image

![image](https://github.com/user-attachments/assets/5852eb49-03cb-4eb3-9133-8f50c68fa0d1)

sudhakara11 commented:
Dear @JyotiSharma0711 @sourabhsaini11 

Thank you for the feedback - We have fixed the type mistake and we will show the fix in the QA demo.

Would you please approve the logs and request for the QA demo?

Regards,
Sudhakar G

sudhakara11 commented:
Dear @JyotiSharma0711 

We have fixed the issues and submitted the logs in the same PR.

Kindly review and approve the logs.

https://github.com/ONDC-Official/verification-logs/pull/1908

Regards,
Sudhakar G

sourabhsaini11 commented:
@sudhakara11
# API Issues and Required Fixes  

## Flow 1  

### on_search  
In **tags**: CONTACT_INFO, GRO_ADDRESS are missing.  


### on_confirm  
fulfillments.state code should be **SANCTIONED** instead of **INITIATED**.  
In **SETTLEMENT_TERMS**, DELAY_INTEREST and SETTLEMENT_TYPE are missing.  

## Flow 2  

### on_update_2  
IDs and **amount** should not be the same as the API contract.  

### on_update_3  
payments[0].params, time, url are missing.  
PRE_PAYMENT_CHARGE should have an additional sub-tag in quote.breakup.  

### on_update_4  
Why is **URL** present in the object at payments[last index]?  

## Flow 3  

### General Issue  
In any of the on_update events, **payment URL is missing**.  
How are **pre_part_payment**, **missed_emi**, and **foreclosure** getting paid without a payment URL?  

### on_update_3  
payments[0].params, time, url are missing.  

## Flow 4  

### on_update_4  
payments.url is missing in on_update.

sudhakara11 commented:
Dear @sourabhsaini11 @JyotiSharma0711 @Antriksh 

We have verified the comments and fixed flow1 GRO_ADDRESS issues and submitted the logs.

https://github.com/ONDC-Official/verification-logs/pull/1908

For remaining flows below are the clarifications 

### Flow 1
**on_search**
In tags: CONTACT_INFO, GRO_ADDRESS are missing.
_**NP: The issues has been fixed and submitted flow1 logs**_

**on_confirm**
fulfillments.state code should be SANCTIONED instead of INITIATED.

_** NP:  P2P Lender side - immediately lender is not sanctioned - first it will be initiated - lender will do additional checks as per of the under writings - then in the unsolicited call - it will be sanctioned**_

In SETTLEMENT_TERMS, DELAY_INTEREST and SETTLEMENT_TYPE are missing.

_**NP: The details are available in the submitted logs**_

### Flow 2
**on_update_2**
IDs and amount should not be the same as the API contract.

_**NP: The details as per the contract**_

**on_update_3**
payments[0].params, time, url are missing.
PRE_PAYMENT_CHARGE should have an additional sub-tag in quote.breakup.

_**NP: PRE_PART_PAYMENT is having all the details but in different index**_

**on_update_4**
Why is URL present in the object at payments[last index]?

_**NP: URL is not present in the json**_

### Flow 3
**General Issue**
In any of the on_update events, payment URL is missing.
How are pre_part_payment, missed_emi, and foreclosure getting paid without a payment URL?

_**NP: payment URLs are available in the on_update_3 for pre_part_payment and foreclosure 
missed_emi payment URLs would be available in on_update_4 - the submitted has these details kindly review**_ 

on_update_3
payments[0].params, time, url are missing.

_**NP:missed_emi flow - on_update_3 wont be having url**_

### Flow 4
**on_update_4**
payments.url is missing in on_update.

_**NP: Foreclouse on_update_4 already payment has been done, so url wont be avaialble**_

Regards,
Sudhakar G

sourabhsaini11 commented:
> on_update_3

Noted @sudhakara11 .
Kindly get these issues cleared as well
## Flow 2

### on_update_4
**PRE_PAYMENT_CHARGE** should have an additional sub-tag in quote.breakup.

## Flow 3
### on_update_3
**payments.url** is missing for payments object with time.label **MISSED_EMI_PAYMENT**

### on_update_4
duplicate call as on_update_3 found 


### on_update_5
**OUTSTANDING_PRINCIPAL** price mismatch found b/w on_update_5 & on_update_4, in quote.breakup.
**LATE_FEE_AMOUNT** should have an additional sub-tag in quote.breakup.


## Flow 4

### on_update_4
**OUTSTANDING_PRINCIPAL & FORECLOSUER_CHARGES** price mismatch found b/w on_update_4 & on_update_3, in quote.breakup.

sudhakara11 commented:
Dear @sourabhsaini11 @JyotiSharma0711 

We have reviewed the observations and below are the clarifications for the same

### Flow 2
**on_update_4**
PRE_PAYMENT_CHARGE should have an additional sub-tag in quote.breakup.

**_NP: PRE_PAYMENT_CHARGE would be available in on_update_3, as it has to be pay, once the payment is done - on_update_4 would be called and PRE_PAYMENT_CHARGE wont be available - we have referred the verified logs (pramaan)_**

### Flow 3
**on_update_3**
payments.url is missing for payments object with time.label MISSED_EMI_PAYMENT

_**NP: At this step the it will be marked as missed_emi, so the URL will be sent in next on_update_4 with same label**_ 
**on_update_4**
duplicate call as on_update_3 found

_**NP: on_update_4 will be used for sending the missed_emi_payment URL**_

**on_update_5**
OUTSTANDING_PRINCIPAL price mismatch found b/w on_update_5 & on_update_4, 

_**NP: on_update_4 the missed_emi_payment has been done and this payment is reflected on_updated_5 OUTSTANDING_PRINCIPAL amount.**_

in quote.breakup.
LATE_FEE_AMOUNT should have an additional sub-tag in quote.breakup.

_**NP: As payment is done in on_update_4 the LATE_FEE_AMOUNT exists, where as on_update_5 not required to have LATE_FEE_AMOUNT ( as per the pramaan)**_


### Flow 4
**on_update_4**
OUTSTANDING_PRINCIPAL & FORECLOSUER_CHARGES price mismatch found b/w on_update_4 & on_update_3, in quote.breakup.

_**NP: on_update_3 the foreclosure has been done and this payment is reflected on_updated_4 OUTSTANDING_PRINCIPAL amount.**_


Regards,
Sudhakara G

sourabhsaini11 commented:
@sudhakara11
Flow 2
NP: PRE_PAYMENT_CHARGE would be available in on_update_3, as it has to be pay, once the payment is done - on_update_4 would be called and PRE_PAYMENT_CHARGE wont be available - we have referred the verified logs (pramaan)
- this needs to be present, check api doc for 2nd on_update for PRE_PAYMENT

Flow 3
NP: At this step the it will be marked as missed_emi, so the URL will be sent in next on_update_4 with same label
NP: on_update_4 will be used for sending the missed_emi_payment URL
NP: on_update_4 the missed_emi_payment has been done and this payment is reflected on_updated_5 OUTSTANDING_PRINCIPAL amount.
NP: As payment is done in on_update_4 the LATE_FEE_AMOUNT exists, where as on_update_5 not required to have LATE_FEE_AMOUNT ( as per the pramaan)
- a single on_update should be sent containing MISSED_EMI_PAYMENT payment obj with payment.url in it, post payment url attribute should be removed with status changing to PAID
- OUTSTANDING_PRINCIPAL should remain same b/w both on_updates
- LATE_FEE_AMOUNT this needs to be present as well

Flow 4
NP: on_update_3 the foreclosure has been done and this payment is reflected on_updated_4 OUTSTANDING_PRINCIPAL amount. FORECLOSUER_CHARGES not required to have as already payment has been done.
- OUTSTANDING_PRINCIPAL & FORECLOSUER_CHARGES should remain same b/w both on_updates

sudhakara11 commented:
Dear @sourabhsaini11 @JyotiSharma0711  @Antriksh

Thank you for the quick response - We have fixed the comments and below are couple of observations.

### Flow 2
NP: PRE_PAYMENT_CHARGE would be available in on_update_3, as it has to be pay, once the payment is done - on_update_4 would be called and PRE_PAYMENT_CHARGE wont be available - we have referred the verified logs (pramaan)

this needs to be present, check api doc for 2nd on_update for PRE_PAYMENT

_**NP: Fixed**_

### Flow 3
NP: At this step the it will be marked as missed_emi, so the URL will be sent in next on_update_4 with same label
NP: on_update_4 will be used for sending the missed_emi_payment URL
NP: on_update_4 the missed_emi_payment has been done and this payment is reflected on_updated_5 OUTSTANDING_PRINCIPAL amount.
NP: As payment is done in on_update_4 the LATE_FEE_AMOUNT exists, where as on_update_5 not required to have LATE_FEE_AMOUNT ( as per the pramaan)


- a single on_update should be sent containing MISSED_EMI_PAYMENT payment obj with payment.url in it, post payment url attribute should be removed with status changing to PAID
- OUTSTANDING_PRINCIPAL should remain same b/w both on_updates
- LATE_FEE_AMOUNT this needs to be present as well

**_NP: Fixed 1 and 3 points_**

### Flow 4
NP: on_update_3 the foreclosure has been done and this payment is reflected on_updated_4 OUTSTANDING_PRINCIPAL amount. FORECLOSUER_CHARGES not required to have as already payment has been done.

OUTSTANDING_PRINCIPAL & FORECLOSUER_CHARGES should remain same b/w both on_updates

**_NP: FORECLOSURE_CHARGES added in all on_update calls._**

Regarding **OUTSTANDING_PRINCIPAL** we have verified with our lender - 

### 1.Example of Missed EMI Payment (with Updated Repayment Plan)

**Scenario:**

- Loan Amount: ₹5,00,000
- Interest Rate: 12% per annum
- Tenure: 24 months
- EMI: ₹23,537 (Principal + Interest)
- Due Date: 5th of every month
- Borrower misses the EMI for March 5th

<img width="569" alt="image" src="https://github.com/user-attachments/assets/c0dd362d-447f-436c-bbd0-1f77bacc068a" />

Here we are paying the missed emi payment so the OUTSTANDING_PRINCIPAL amount is reduced.


### 2. Foreclosure Example

Foreclosure occurs when the borrower repays the entire outstanding loan before the original tenure ends.

### Scenario:

- Loan Amount: ₹5,00,000
- Tenure: 24 months
- EMI Paid for 12 months
- Outstanding Principal: ₹2,80,000
- Foreclosure Charges: 2% of outstanding amount (₹5,600)
- Total Amount to Close Loan: ₹2,85,600

**Case: Borrower Pays ₹2,85,600 in Month 13**

- The loan is closed early, avoiding future interest costs.
- The lender updates the OUTSTANDING_PRINCIPAL to ₹0 in the next ONDC update call.
- The borrower gets a loan closure letter confirming completion.

Regards,
Sudhakar G

sourabhsaini11 commented:
@sudhakara11, transactional-logs cleared

sudhakara11 commented:
Dear @sourabhsaini11 @JyotiSharma0711 

Thank you for the continuous support on the log review.

We have submitted IGM logs as well, Kindly review and provide the feedback.

Regards,
Sudhakar G

amitsinha07 commented:
@sudhakara11 

## IGM feedbacks

`on_issue_status`  
`issue/created_at` is not matching with `issue_open`'s `issue/created_at`

sudhakara11 commented:
Dear @amitsinha07 

Thank you for the quick support.

We have fixed the issue and submitted the logs on the same PR.

Kindly review and confirm.

Regards,
Sudhakar G

amitsinha07 commented:
@sudhakara11 

### IGM feedbacks
`on_issue_1`  
`issue/created_at` is not matching with `issue_open`'s `issue/created_at`

amitsinha07 commented:
@sudhakara11 
Your logs are approved for IGM 1.0.0

sudhakara11 commented:
@amitsinha07  @JyotiSharma0711 

Thank you for the continuous support.

Regards,
Sudhakar G

sudhakara11 commented:
Dear @sourabhsaini11  @amitsinha07 @92shreyansh 

As the transactions/igm logs are approved, kindly merge the MR.

Regards,
Sudhakar G
